controllers/reservation.controllers.js

```diff
@@ -83,47 +81,473 @@ const listReservationByDate = async ({ client, ack, say, body }) => {
+  const isRoomFull = currentReservations.length
+    && currentReservations.length >= currentReservations[0].room.maxCapacity;
+
+  const startOfWeek = moment(date).startOf('isoWeek');
+  const endOfWeek = moment(date).endOf('isoWeek');
+
+  const query = { date: { $gt: startOfWeek, $lt: endOfWeek }, user: user._id };
+  const currentOffice = await Office.findById(user.office);
+  const alreadyHaveReserve = await Reservation.find({
+    date,
+    office: user.office,
+    room: room.value,
+    user: user._id,
+  });
+
+  const weekResevationByUser = await Reservation.find(query);
+
+  if (invalidDate) {
+    errors = {
+      response_action: 'errors',
+      errors: {
+        date_input: 'La fecha no es válida',
+      },
+    };
+  }
+
+  if (!room) {
+    errors = {
+      response_action: 'errors',
+      errors: {
+        site_input: 'Seleccione un lugar',
+      },
+    };
+  }
+
+  if (!frecuency) {
+    errors = {
+      response_action: 'errors',
+      errors: {
+        date_input: 'El frecuencia no es válida',
+      },
+    };
+  }
+
+  if (isRoomFull) {
+    errors = {
+      response_action: 'errors',
+      errors: {
+        site_input: 'No hay lugares disponible en esta sala',
+      },
+    };
+  }
+
+  if (alreadyHaveReserve.length) {
+    errors = {
+      response_action: 'errors',
+      errors: {
+        date_input: 'Ya tienes una reserva para este día',
+      },
+    };
+  }
+
+  if (weekResevationByUser.length > currentOffice.maxVisitsAWeek) {
+    errors = {
+      response_action: 'errors',
+      errors: {
+        date_input: 'Alcanzaste el máximo de reservas esta semana',
+      },
+    };
+  }
+
+  await ack(errors);
+
+  let message;
+
+  const hasErrors = Object.keys(errors).length;
+
+  if (!hasErrors) {
+    let reservation;
+    if (frecuency.value === 'day') {
+      reservation = await Reservation.create({
+        date,
+        user: user._id,
+        team: user.team,
+        office: user.office,
+        room: room.value,
+      });
+    }
+
+    if (frecuency.value === 'week') {
+      const start = moment(date).startOf('day');
+      const end = moment(date).endOf('isoWeek');
+
+      const days = [];
+      let day = start;
+
+      while (day <= end) {
+        days.push(day.toDate());
+        day = day.clone().add(1, 'd');
+      }
+
+      reservation = days.map(
+        async (day) => await Reservation.create({
+          date: day,
+          user: user._id,
+          team: user.team,
+          office: user.office,
+          room: room.value,
+        }),
+      );
+    }
+
+    if (frecuency.value === 'month') {
+      const start = moment(date).startOf('day');
+      const end = moment(date).endOf('month');
+      const days = [];
+      let day = start;
+
+      while (day <= end) {
+        days.push(day.toDate());
+        day = day.clone().add(1, 'd');
+      }
+
+      reservation = days.map(
+        async (day) => await Reservation.create({
```

Comment:
would it make sense to do some sort of batching to avoid doing a create per each element?